lib/Runtime/Base/CompactCounters.h

```diff
@@ -0,0 +1,269 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+#pragma once
+
+#pragma warning(push)
+#pragma warning(disable:6200) // C6200: Index is out of valid index range, compiler complains here we use variable length array
+
+namespace Js
+{
+    template<class T, typename CountT = T::CounterFields>
+    struct Counter
+    {
+
+        volatile uint8 fieldSize;
+        union {
+            uint8 u8fields[CountT::Max];
+            int8 i8fields[CountT::Max];
+            WriteBarrierPtr<uint16> u16Fileds;
+            WriteBarrierPtr<uint32> u32Fileds;
+            WriteBarrierPtr<int16> i16Fileds;
+            WriteBarrierPtr<int32> i32Fileds;
+        };
+
+
+        Counter()
+            :fieldSize(1)
+        {
+            memset(u8fields, 0, (uint8)CountT::Max);
+        }
+
+        void AllocCounters(T* host, uint8 newSize);
+
+        uint32 Get(CountT typeEnum) const
+        {
+            uint8 type = static_cast<uint8>(typeEnum);
+            uint8 localFieldSize = fieldSize;
+            uint32 value = 0;
+            if (localFieldSize == 1)
+            {
+                value = this->u8fields[type];
+            }
+            else if (localFieldSize == 2)
+            {
+                value = this->u16Fileds[type];
```

Comment:
Since we didn't take a lock here, background thread could potentially read old value from u8fields here even after we switch to u16fields.   May guess it that it is probably ok, because we didn't have a lock before, so the background thread either is just access field that doesn't change after byte code gen, or some count that doesn't matter if we missed some.     But please verify that is true.

Please add a comment here too, so we won't think it is a bug in the future.